bach/bach/dataframe.py

```diff
@@ -1545,25 +1543,65 @@ def groupby(
             group_by = GroupingSet(
                 [GroupBy(group_by_columns=df._partition_by_series(b)) for b in by]
             )
-        elif isinstance(by, list) and len([b for b in by if isinstance(b, (tuple, list))]) > 0:
+            return DataFrame._groupby_to_frame(df, group_by)
+
+        if isinstance(by, list) and len([b for b in by if isinstance(b, (tuple, list))]) > 0:
+            if not is_postgres(self.engine):
+                raise DatabaseNotSupportedException(
+                    self.engine,
+                    f'GroupingLists are not supported for this SQL dialect. Only grouping by one or more '
+                    f'Series is supported. To use this functionality: specify a Series, a list of Series, a '
+                    f'Series name, or a list of Series names. Dialect: {self.engine.name}')
             group_by = GroupingList(
                 [GroupBy(group_by_columns=df._partition_by_series(b)) for b in by])
-        else:
-            by_mypy = cast(Union[str, 'Series',
-                                 List[DataFrame._GroupBySingleType], None], by)
-            group_by = GroupBy(group_by_columns=df._partition_by_series(by_mypy))
+            return DataFrame._groupby_to_frame(df, group_by)
 
+        # This is the normal group-by case
+        by_mypy = cast(Union[str, 'Series',
+                             List[DataFrame._GroupBySingleType], None], by)
+        group_by_columns = df._partition_by_series(by_mypy)
+
+        group_by_expressions = any(
+            gbc.expression != Expression.column_reference(gbc.name) for gbc in group_by_columns
+        )
+        if is_bigquery(self.engine) and group_by_expressions:
+            # BigQuery allows grouping by an expression (other than just a column-reference), but then that
+            # expression cannot be in the select (unless all columns that are in the expression are also
+            # grouped on). However, we always include all expressions that are grouped on in the result, as
+            # that is the new index. So this presents a problem.
+            # To prevent problems with this we add the group-by expressions to the frame, materialize them,
+            # and then create the group-by.
+            # TODO: situations with more than one expression that refers the same column (for all databases)
+            #   e.g. df.groupby([df.x >= 0, df.x == 0])
+            #   See https://github.com/objectiv/objectiv-analytics/issues/616
+            df_new = df.copy()
+            gbc_names: List[str] = [gbc.name for gbc in group_by_columns]
+            for gbc in group_by_columns:
+                df_new[gbc.name] = gbc
+            df_new = df_new.materialize(node_name='bq_materialize_before_groupby')
+            group_by_columns = [df_new[gbc_name] for gbc_name in gbc_names]
+            group_by = GroupBy(group_by_columns=group_by_columns)
+            return DataFrame._groupby_to_frame(df_new, group_by)
```

Comment:
maybe? it is already iterating over `group_by_columns` and `_partition_by_series` already does the job of getting group by columns, so just consistency
```suggestion
            gbc_names = []
            for gbc in group_by_columns:
                gbc_names.append(gbc.name)
                df_new[gbc.name] = gbc

            df_new = df_new.materialize(node_name='bq_materialize_before_groupby')
            group_by_columns = df_new._partition_by_series(gbc_names)
            group_by = GroupBy(group_by_columns=group_by_columns)
            return DataFrame._groupby_to_frame(df_new, group_by)
```

Reply:
Not really an improvement imho. Because if I read this now I have to click through to `_partition_by_series()` and understand what that does to understand this fragment.

However, I don't care about this very strongly and since you think it's better, I will apply this change too